### DISCUSSION TITLE
why zod doesn't make no empty string by default for required fields?

### DISCUSSION THREAD

**Discussion by Chandratand** [colinhacks/zod#2874] [General]
I am using react hook form and zod as a validator resolver.

```ts
export const SignInValidator = z.object( {
    email: z.string().email( "Invalid email address" ),
    password: z.string(),
} )

const form = useForm( {
    resolver: zodResolver( SignInValidator ),
    defaultValues: {
        email: "",
        password: "",
    },
} )
```
  
SignInValidator is passes even if the password is an empty string. the default value is empty string, if I don't create an empty string at the default value, a Warning will occur: A component changed an uncontrolled input to a controlled one. This is likely caused by changing a value from undefined to a defined value, which should not happen. Decide between using controlled or uncontrolled input elements during the lifetime of the component.

why zod doesn't make noEmptystring by default for required fields?

**Comment by JacobWeisenburger** [comment]
A required field just means that the value must exist and must not be `undefined`. An empty string `''` does in fact exist and it is not undefined, so it is valid. I think what you are looking for is `z.string().min( 1 )`. This will require that the string is at least 1 character long.

If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger

**Comment by Chandratand** [comment]
Actually I know I can use min(1, "Required") to validate it. it's just that I'm confused why it's not made by default because in case the default value of an input is an empty string and the required validation in zod allows it to pass.

if we don't set the default value of an input as an empty string, this warning will occur: A component changed an uncontrolled input to a controlled one. This is likely caused by changing a value from undefined to a defined value, which should not happen. Decide between using controlled or uncontrolled input elements during the lifetime of the component.

**Comment by JacobWeisenburger** [comment]
Zod's purpose was not to help with React Hook Form. So it's not Zod's job to cater to how another library works. Zod's purpose was to be TypeScript compatible. So I don't imagine Zod will ever consider an empty string to be the same as `undefined`. They are fundamentally 2 different values.

If you are looking for an explanation of the reasoning, then I'll move this to discussions.

If you are wanting Zod to change how it handles this, then I'll leave this as an open issue.

Which would you like?

**Comment by Chandratand** [comment]
Actually, this applies not only to react-hook forms but also to controlled inputs (inputs with values set via state).

Yes, I understand your explanation. And I also already know the solution by using min(1, "Required"). I just felt that it would be more effective if done by default.

It's fine if you want to keep this as an open issue or move it to the discussion. i am fine with either one.

**Comment by awacode21** [comment]
how do you handle a generic required rule for several types of inputs like text, number, radio?

**Reply by BrendanC23** [reply]
If you don't know the type of data, you can build a schema that incorporates multiple possibilities into a union.

```ts
const schema = z.object({
    field: z.union(z.string().min(1), z.number(), z.string().array()),
}).superRefine((data, ctx) => {
    // Any other validation
});
```

**Comment by vitorschelb** [comment]
Im using ShadCN, Hook Forms and Zod.
Select in ShadCN accepts empty string as valid value, doesnt show placeholder.
Zod doesnt show custom error message for undefined and null values.
This solved the problem:

  fieldName: z
    .string()
    .nullable()
    .refine((val) => !!val, {
      message: "Custom Message",
    }),

**Comment by pret-a-porter** [comment]
Now you can use `z.string().nonempty()`
